open-apparel-registry/open-apparel-registry
# Update sectors in FacilityIndex after upload

## Overview

We previously didn't update the entire FacilityIndex
after a FacilityListItem was saved, only the custom
text. In most cases, the post_save for the FacilityMatch
covered the needed updates to the index. However,
in certain cases, the FacilityListItem for a new
facility is saved after the FacilityMatch is saved.
In these cases, when the Match.post_save updates the
FacilityIndex, the FacilityListItem's status is not yet
MATCHED or CONFIRMED_MATCH, so fields such as sector that
are only added to the index if a FacilityListItem with one of
those statuses exists are excluded.

Adding indexing to the FacilityListItem.post_save when the
item is in a valid status results in the sectors being
added to the index as expected.

Connects #1990

## Demo

<img width="346" alt="Screen Shot 2022-07-27 at 2 53 17 PM" src="https://user-images.githubusercontent.com/21046714/181508621-d8049499-d8f3-4406-abe4-45d243c108c2.png">

## Notes

This PR won't fix existing FacilityIndex rows with missing sector fields. We will need to manually run the `index_facilities` management command in order to resolve existing issues. I recommend re-indexing all facilities, since all facilities have the sector field, to ensure we've handled all relevant cases. 

## Testing Instructions

On `develop`:
* Run `./scripts/resetdb` to make sure no facilities will match with the facilities in [TestSector.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/9209711/TestSector.csv)
* Run `./scripts/server`
* Upload TestSector.csv
* Run `./tools/batch_process 16`
* Visit the facility search, refresh the page, and note that 'Food and Beverages' and 'Agriculture and Farming' haven't been added to the dropdown

On this branch: 
* Run `./scripts/resetdb` to make sure no facilities will match with the facilities in TestSector.csv
* Run `./scripts/server`
* Upload TestSector.csv and note the list id
* Run `./tools/batch_process 16`
* Visit the facility search, refresh the page, and note that 'Food and Beverages' and 'Agriculture and Farming' have been added to the dropdown

## Checklist
